systemcontracts: apply baseFee from policy contract for burning

chenquanyu opened the pull request:
Close #163
Add the `baseFee` reading from policy contract for `eip1559.CalcBaseFee()`. 

After this PR the `BaseFee`  field in block header will be the `baseFee` in policy contract, so this will cause hard fork for current testnet. I have added `neoxburnTime` field in genesis config, `neoxburnTime` should be set to a future timestamp and the nodes need to be re-init if we want to upgrade the testnet.

I have tested the upgrade process in privnet seven node, the `BaseFee` in block header changes after the `neoxburnTime` timestamp and transaction need to pay at least `baseFee` + `minGasTipCap` in policy contract.

steven1227 commented:
Ready for reviewing and testing.

AnnaShaleva commented on consensus/dbft/dbft.go:
```diff
@@ -903,6 +903,12 @@ func (c *DBFT) verifyCascadingFields(chain consensus.ChainHeaderReader, header *
+	if err != nil {
+		return fmt.Errorf("failed to get state: err %w, root %s", err, parent.Root)
+	}
+
```
We can't do this, it's too invasive and it's not a proper place for this verification. `verifyCascadingFields` belongs to the common engine verification method (`VerifyHeader`), and should not access the node state. Consider running this code on a light node.

dBFT doesn't has its own snapshot, its verification mechanism is different from Clique. Ordinary nodes should just ensure that this block is signed by proper consensus nodes (this check is already present). And consensus nodes must ensure the base fee correctness during PrepareRequest state-dependent verification (inside the `WithVerifyBlock` callback).

chenquanyu replied:
OK, I will move this check to `WithVerifyBlock`.

AnnaShaleva replied:
> will move this check to WithVerifyBlock.

I think that this check should be moved to state-dependent transition verification. With this change no additional code is required in `WithVerifyBlock`.

AnnaShaleva commented on consensus/misc/eip1559/eip1559.go:
```diff
@@ -54,12 +56,20 @@ func VerifyEIP1559Header(config *params.ChainConfig, parent, header *types.Heade
+			return baseFee
+		}
+	}
+
```
Also can't do this here since `CalcBaseFee` should be state-independent. We need to move it inside the state-dependent verification.

chenquanyu replied:
But now we get `baseFee` from Policy contract, we need `state` to verify the `baseFee`. Adding a method to get `baseFee` from `state` is no different from the current implementation.

AnnaShaleva replied:
> we need state to verify the baseFee. 

That's why this check can't be present at `VerifyEIP1559Header`. Headers verification should be state-independent.

AnnaShaleva reviewed (commented):
I think we need to rework the code, there's a lot of places where the block state is being fetched although I'd say it shouldn't.

AnnaShaleva commented on core/blockchain.go:
```diff
@@ -2471,6 +2472,12 @@ func (bc *BlockChain) ProcessState(block *types.Block) (*state.StateDB, types.Re
+	if block.BaseFee().Cmp(baseFee) != 0 {
+		return nil, nil, nil, 0, fmt.Errorf("failed to verify policy baseFee, expected: %v, current: %v", baseFee, block.BaseFee())
+	}
+
```
I think it'll be better to move this check to `(st *StateTransition) preCheck()` since this check is needed not only for dBFT blocks processing but also in the process of normal block persist cycle.

chenquanyu replied:
`(st *StateTransition) preCheck()` is for message level check and only dBFT blocks need this check because others are check in `VerifyEIP1559Header`.

roman-khimov replied:
I don't think we have any other blocks but dBFT in Bane. And looks like this change follows the practice of "check during consensus, trust in use" for dBFT header fields (consider state-dependent `NextConsensus`, for example).

roman-khimov commented on consensus/clique/clique.go:
```diff
@@ -357,7 +357,7 @@ func (c *Clique) verifyCascadingFields(chain consensus.ChainHeaderReader, header
 			return err
 		}
-	} else if err := eip1559.VerifyEIP1559Header(chain.Config(), parent, header); err != nil {
+	} else if err := eip1559.VerifyEIP1559Header(chain.Config(), parent, chain.CurrentHeader()); err != nil {
```
Why? I thought we're checking `header` specifically here.

chenquanyu replied:
Sorry, here is unnecessary change.

roman-khimov commented on consensus/misc/eip1559/eip1559.go:
```diff
@@ -54,12 +56,20 @@ func VerifyEIP1559Header(config *params.ChainConfig, parent, header *types.Heade
 	}
 
+	// Read baseFee from policy if baseFee is set in policy contract
+	if state != nil {
```
I see it can be `nil` in your code, but I really doubt it should ever be this way.

chenquanyu replied:
This file is changed in latest code, please refresh.

roman-khimov commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -1281,7 +1281,14 @@ func (pool *LegacyPool) runReorg(done chan struct{}, reset *txpoolResetRequest,
+					state, err = pool.chain.StateAt(types.EmptyRootHash)
+				}
+				if err != nil {
+					log.Error("Get state error", "err", err, "root", reset.newHead.Root)
```
Is it safe to continue here?

chenquanyu replied:
Sorry, I check the code again and find that the state is saved in `reset(oldHead, newHead *types.Header)`, so we can just use `pool.currentState` here.

roman-khimov commented on privnet/seven/genesis_privnet.json:
```diff
@@ -15,6 +15,7 @@
     "arrowGlacierBlock": 0,
     "grayGlacierBlock": 0,
     "shanghaiTime": 0,
+    "neoxburnTime": 0,
```
I hope it's testnet-only.

chenquanyu replied:
It's for t3 testnet hard fork, but for mainnet we should also include this to apply `baseFee` change.

steven1227 replied:
It's better to keep this, same as `shanghaiTime` above. Can change a better name

AnnaShaleva replied:
> I hope it's testnet-only.

For mainnet it will be enabled starting from 0, but we still need to include this entry to the config (given the fact that the current testnet should be supported with non-zero `neoxburnTime`). @roman-khimov, are you against of adding this entry to the mainnet config?

roman-khimov reviewed (commented):
Looks legit otherwise.

chenquanyu committed:
systemcontracts: apply baseFee from policy contract for burning
Close #163

AnnaShaleva commented on consensus/misc/eip1559/eip1559.go:
```diff
@@ -93,3 +95,32 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header) *big.Int {
+	}
+	// Verifing the baseFee is moved to VerifyBlock.
+	return nil
+}
```
I don't quite like the massive code duplication between `VerifyEIP1559HeaderDBFT` and `VerifyEIP1559Header`. Base fee verification and calculation is bound to the `IsNeoxburn` hardfork, thus can we keep the `VerifyEIP1559Header` and just move the last part of it under `if config.IsNeoxburn` condition? Like this:
```go
	...
	// Verify the baseFee is correct based on the parent header. If neoxburn hardfork enabled, then
	// base fee verificaiton is state-dependent and performed by state-dependent dBFT block verification.
	if config.IsNeoxburn {
		return nil
	}
	expectedBaseFee := CalcBaseFee(config, parent)
	if header.BaseFee.Cmp(expectedBaseFee) != 0 {
		return fmt.Errorf("invalid baseFee: have %s, want %s, parentBaseFee %s, parentGasUsed %d",
			header.BaseFee, expectedBaseFee, parent.BaseFee, parent.GasUsed)
	}
	...
```

AnnaShaleva replied:
Implemented in #230.

AnnaShaleva commented on consensus/misc/eip1559/eip1559.go:
```diff
@@ -93,3 +95,32 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header) *big.Int {
+		return CalcBaseFee(config, parent)
+	}
+	return state.GetState(systemcontracts.PolicyProxyHash, systemcontracts.GetBaseFeeStateHash()).Big()
+}
```
We're replacing all the usages of `CalcBaseFee` by `CalcBaseFeeDBFT`. The better way to me is to remove `CalcBaseFeeDBFT` function, add `state *state.StateDB` argument to the original `CalcBaseFee` and move this `config.IsNeoxburn`-dependent logic under `CalcBaseFee`. This approach allows us to easily rebase onto fresh Geth changes without unexpected missings. It also allows to ensure that all usages of `CalcBaseFee` are updated properly.

chenquanyu replied:
I added `state *state.StateDB` argument before, but changed for your comment https://github.com/bane-labs/go-ethereum/pull/166#discussion_r1576134355

AnnaShaleva replied:
Oh, I see, my bad. But the current way also doesn't seem to be right, we replacing all the usages of CalcBaseFee by CalcBaseFeeDBFT anyway. Let me take more detailed look at it.

AnnaShaleva replied:
In #230 I tried to reduce the number of usages of CalcBaseFeeDBFT, so there's not a lot of them left. One thing that prevent us from CalcBaseFee and CalcBaseFeeDBFT unification is testing code and `ChainMaker`, they still use an old version (CalcBaseFee) that is not capable of returning Policy-based value. We need to decide if we'd like to do something about it.

AnnaShaleva commented on core/blockchain.go:
```diff
@@ -2471,6 +2472,12 @@ func (bc *BlockChain) ProcessState(block *types.Block) (*state.StateDB, types.Re
+	if block.BaseFee().Cmp(baseFee) != 0 {
+		return nil, nil, nil, 0, fmt.Errorf("failed to verify policy baseFee, expected: %v, current: %v", baseFee, block.BaseFee())
+	}
+
```
To me, it's more a part of `(bc *BlockChain) VerifyBlock`, because this check is not related to the state processor itself.

Also, dBFT uses ProcessState during PrepareRequest creation where this verification code shouldn't be executed.

AnnaShaleva replied:
Also we do retrieve parent state one more time a few lines below during state transition. We need to pay more attention to our computational resources, otherwise we'll significantly slow down the node.

chenquanyu replied:
I add the verify here because the `statedb` is retrieved here, we will not  retrieve parent state one more time.

AnnaShaleva replied:
> we will not retrieve parent state one more time.

That's true, seems that I misread the code, so the second problem is not actually a problem. However, the first problem still exist. The thing is: this check is not a part of state processing. When we do state processing, we assume that header has already passed verification and is valid. So I'd place this check to `VerifyBlock`, after body validation but before state processing. We can get the state to verify it and then pass the same state to `bc.ProcessState` as a second argument. At the same time, `bc.ProcessState` need to check whether provided state is nil, and if it is, then retrieve state by itself.

chenquanyu replied:
OK, got it.

AnnaShaleva commented on core/state_transition.go:
```diff
@@ -331,6 +324,16 @@ func (st *StateTransition) preCheck() error {
+				// For LegacyTx, GasFeeCap and GasPrice are equal, so checking GasTipCap and GasFeeCap is enough
+				var minGasTipCap = st.state.GetState(systemcontracts.PolicyProxyHash, systemcontracts.GetMinGasTipCapStateHash()).Big()
+				if cmath.BigMin(msg.GasTipCap, new(big.Int).Sub(msg.GasFeeCap, st.evm.Context.BaseFee)).Cmp(minGasTipCap) < 0 {
+					return fmt.Errorf("%w: address %v, gastipcap %v, gasfeecap %v", ErrUnderpriced,
```
We need to add `minGasTipCap` required by Policy to the error message.

chenquanyu replied:
Fixed.

AnnaShaleva commented on core/txpool/validation.go:
```diff
@@ -203,9 +204,11 @@ func ValidateTransactionWithState(tx *types.Transaction, signer types.Signer, op
-		return fmt.Errorf("%w: policy needed %v, tip permitted %v", ErrUnderpriced, minGasTipCap.Big(), tx.GasTipCap())
+	var baseFee = opts.State.GetState(systemcontracts.PolicyProxyHash, systemcontracts.GetBaseFeeStateHash()).Big()
+	if math.BigMin(tx.GasTipCap(), new(big.Int).Sub(tx.GasFeeCap(), baseFee)).Cmp(minGasTipCap.Big()) < 0 {
+		return fmt.Errorf("%w: policy needed %v, gastipcap %v, gasfeecap %v ", ErrUnderpriced, minGasTipCap.Big(), tx.GasTipCap(), tx.GasFeeCap())
```
Need to add `baseFee` value to the error message, otherwise it may be not clear why transaction is declined.

chenquanyu replied:
Fixed.

AnnaShaleva commented on eth/gasprice/feehistory.go:
```diff
@@ -83,7 +83,11 @@ func (oracle *Oracle) processBlock(bf *blockFees, percentiles []float64) {
+		if err != nil {
+			log.Error("Failed to get state", "err", err, "header number", bf.blockNumber)
+		}
+		bf.results.nextBaseFee = eip1559.CalcBaseFeeDBFT(chainconfig, bf.header, state)
```
We are not supposed to use node state inside Oracle, and it's a design problem. Have you tried to test `FeeHistory` call on the light-weighted (state-less) nodes? It likely won't work, because here we refer to the node state. Correct me if I'm wrong, please.

We need a BaseFee calculation scheme that allows to compute next block base fee without referring to the parent's state. Ideally, it would be nice to have have something similar to NextConsensus logic: it's included in the parent's block, and we don't need an access to state to verify the next block, we only need a parent's header for it.

AnnaShaleva replied:
@chenquanyu, @chenquanyu, @roman-khimov this is the main issue I'd like to clarify (and may be fix).

>And looks like this change follows the practice of "check during consensus, trust in use" for dBFT header fields (consider state-dependent NextConsensus, for example).

That's true, so the only component that should *compute* the base fee from the state is a consensus engine. The problem is I see this state-dependent code in some other components like Oracle or Eth API which is not quite correct to me.

roman-khimov replied:
IIUC there are cases when we need base fee for the next block outside consensus, like test executions.

roman-khimov replied:
If you're proposing something like `NextBaseFee` added to the block header, it might be interesting.

AnnaShaleva replied:
> IIUC there are cases when we need base fee for the next block outside consensus, like test executions.

But the problem is not all of these cases are powered to access the blockchain state. Currently light-weighted node compatibility is not tested at all, and I'm just afraid that we'll break it completely.

> something like NextBaseFee added to the block header,

Yes, something like that, because I don't see another option to solve the mentioned problem.

chenquanyu replied:
> If you're proposing something like `NextBaseFee` added to the block header, it might be interesting.

Do you mean adding the `NextBaseFee` field to the block header, or adding `NextBaseFee` to the `Extra` field of the block header? Currently we have not added fields to the block header, which may affect other tools to parse the block.

txhsl replied:
Why don't we use `header.BaseFee` here?
https://github.com/bane-labs/go-ethereum/blob/fe076a097f83605e32961846309a0ecc8c8b141a/eth/gasprice/feehistory.go#L301-L306
What about change the length of `basefee` to `blocks`, and try to append the latest one in `abi_backend`.

AnnaShaleva replied:
> What about change the length of basefee to blocks

I'm afraid we can't. It's user-facing behaviour and this extra value is documented to be returned, dependent applications rely on it.

chenquanyu replied:
Light client has been removed from geth, so it's OK to get `state` from `oracle.backend` now.

AnnaShaleva commented on miner/worker.go:
```diff
@@ -951,7 +951,12 @@ func (w *worker) prepareWork(genParams *generateParams) (*environment, error) {
+			return nil, err
+		}
+		// Calculate baseFee with config
+		header.BaseFee = eip1559.CalcBaseFeeDBFT(w.chainConfig, parent, state)
```
State fetching is a quite resource-consuming operation, we need to try to minimize it and reuse an existing state where possible. Miner has its own state that is created inside the `env, err := w.makeEnv(parent, header, genParams.coinbase)` a few lines below. Can we reuse it? And then update calculated BaseFee value where needed.

chenquanyu replied:
Yes, we can reuse it. Fixed.

AnnaShaleva commented on params/config.go:
```diff
@@ -351,6 +351,7 @@ type ChainConfig struct {
 	CancunTime   *uint64 `json:"cancunTime,omitempty"`   // Cancun switch time (nil = no fork, 0 = already on cancun)
 	PragueTime   *uint64 `json:"pragueTime,omitempty"`   // Prague switch time (nil = no fork, 0 = already on prague)
 	VerkleTime   *uint64 `json:"verkleTime,omitempty"`   // Verkle switch time (nil = no fork, 0 = already on verkle)
+	NeoxburnTime *uint64 `json:"neoxburnTime,omitempty"` // Neoxburn switch time (nil = no fork, 0 = already on neoxburn)
```
Please, rename it to `NeoXBurnTime` for the field and to `neoXBurnTime` for the JSON.

chenquanyu replied:
Done.

AnnaShaleva commented on privnet/seven/genesis_privnet.json:
```diff
@@ -15,6 +15,7 @@
     "arrowGlacierBlock": 0,
     "grayGlacierBlock": 0,
     "shanghaiTime": 0,
+    "neoxburnTime": 0,
```
Please, add the corresponding entry to the Testnet config. Let's schedule this hardfork in ~2 weeks from now and then reschedule if needed. We need to review, test and merge all pending changes by this time.

chenquanyu replied:
Will change the Testnet config when the Testnet is ready to update.

AnnaShaleva commented on cmd/evm/internal/t8ntool/execution.go:
```diff
@@ -156,6 +157,13 @@ func (pre *Prestate) Apply(vmConfig vm.Config, chainConfig *params.ChainConfig,
+		BaseFee:  env.ParentBaseFee,
+		GasUsed:  env.ParentGasUsed,
+		GasLimit: env.ParentGasLimit,
+	}, statedb)
```
Not sure about this change, I need some more time to check it.

AnnaShaleva replied:
`t8ntool` is an environment for Eth developers to simulate transactions/blocks processing. It uses configuration file with *some* contract storage data and *some* previous block data. Usually these data include only those contracts that a developer is interested in.

This change requires Policy contract and its storage items to be added to t8ntool configuration to work properly with our BaseFee-dependent logic. My concern is no one knows about this additional logic except us. I suggest two options:
1. Adjust documentation of t8ntool, add examples of exact values that need to be added to the t8ntool launch configuration and explain why developers ever need to take care of Policy contract storage values. Otherwise we break t8ntool and no one knows what exactly should be put into configuration to make it work.
2. Replace this code with a check that user-configured `pre.Env` value is set. Warn if it's not set and warn about the consequences of it. If it's set, then manually add this value to the state DB (to Policy storage). This code should likely be a part of `MakePreState`. This option also requires documentation adjustment.

The second option looks much more preferable to me. Bane developers do not need to put any Policy-cpecific values into their t8ntool configuration. Everything they need to do is to fill in the `currentBaseFee` field.

Either 1 or 2 requires t8ntool test (run it with some real config and try to process a transaction), we can't merge this change without any tests, otherwise there's a risk to break this tool.

AnnaShaleva replied:
The second option is implemented in #230.

chenquanyu committed:
add log for baseFee and minGasTipCap

chenquanyu committed:
reuse state in worker

chenquanyu committed:
rename neoxburn to neoXBurn

chenquanyu committed:
move basefee check to VerifyBlock

AnnaShaleva committed:
t8ntool: properly handle BaseFee set in the env config

If user sets BaseFee in the t8ntool environment configuration, then
the same value should be present in the Policy storage. Also, ensure
that this value is always set by the user or NeoXBurn-enabled networks.

Ref
https://github.com/bane-labs/go-ethereum/pull/166#discussion_r1629068088,
the second option is implemented.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
consensus: squash VerifyEIP1559Header and VerifyEIP1559HeaderDBFT

There's no need in two different methods since it's easy to miss the
update of the original Geth method. Make the VerifyEIP1559Header depend
on the NeoXBurn hardfork.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
core: adjust the documentation

Make the documentation follow the implementation.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
gasprice: cache BaseFee and MinGasTipCap inside Oracle service

Oracle is capable of working with a full backend only anyway, thus
bringing state dependency doesn't make things worse for this service.
However, retrieving full state DB for every request is too
resource-consuming, we'll end up in a very bad situation if not caching
these values.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
ethapi: add BaseFee cache

Cache BaseFee values for TxPoolAPI and TransactionAPI.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
privnet: enable NeoXBurn hardfork from genesis

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
gasprice: return in case of nil block

This code was accidentally removed in
e427cda61546c741f0372c413dedd25292ba6f59.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
eth: remove useless comment

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
filters: properly process nil current block on pending txs retrieval

It's possible that current block is nil, we have to live with it
somehow. Fix the following failing test:
```
--- FAIL: TestPendingTxFilterFullTx (1.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1c0 pc=0xdc1715]

goroutine 52 [running]:
testing.tRunner.func1.2({0xf66340, 0x19f14a0})
	/usr/local/go/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1634 +0x377
panic({0xf66340?, 0x19f14a0?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
github.com/ethereum/go-ethereum/eth/filters.(*FilterAPI).GetFilterChanges(0xc0020020c0, {0xc000c320f0, 0x22})
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/eth/filters/api.go:468 +0x235
github.com/ethereum/go-ethereum/eth/filters.TestPendingTxFilterFullTx(0xc0006064e0)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/eth/filters/filter_system_test.go:338 +0x826
testing.tRunner(0xc0006064e0, 0x11fc210)
	/usr/local/go/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1742 +0x390
FAIL	github.com/ethereum/go-ethereum/eth/filters	1.201s
?   	github.com/ethereum/go-ethereum/eth/gasestimator	[no test files]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xc061b5]

goroutine 25 [running]:
math/big.(*Int).Set(...)
	/usr/local/go/src/math/big/int.go:98
github.com/ethereum/go-ethereum/eth/gasprice.(*Oracle).suggestTipCapInternal(0xc0002a4500, {0x1084ad0, 0x16de8a0}, 0xc000324288)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/eth/gasprice/gasprice.go:249 +0x1fd5
github.com/ethereum/go-ethereum/eth/gasprice.(*Oracle).processBlock(0xc0002a4500, {0x1084ad0, 0x16de8a0}, 0xc0001d6c40, {0x0, 0x0, 0x0?})
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/eth/gasprice/feehistory.go:85 +0x1af
github.com/ethereum/go-ethereum/eth/gasprice.(*Oracle).FeeHistory.func1()
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/eth/gasprice/feehistory.go:290 +0x605
created by github.com/ethereum/go-ethereum/eth/gasprice.(*Oracle).FeeHistory in goroutine 15
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/eth/gasprice/feehistory.go:259 +0x6c5
FAIL	github.com/ethereum/go-ethereum/eth/gasprice	0.030s

```

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
t8ntool: get back the baseFee setter for London

Get back the code removed in fb84a68095e31b2998316a366faffd88b3445147.
If it's London, then BaseFee should still be calculated based on the
parent's fee. If it's NeoXBurn, then just ensure that BaseFee is set in
the config (this check is already present in the applyNeoXBurnChecks.

Fix the following test:
```
--- FAIL: TestT8n (0.39s)
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/1/alloc.json --input.txs ./testdata/1/txs.json --input.env ./testdata/1/env.json --state.fork Frontier+1346
    test_cmd.go:261: (stderr:1) ERROR(3): failed constructing chain configuration: syntax error, invalid eip number 1346
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/1/alloc.json --input.txs ./testdata/1/txs.json --input.env ./testdata/1/env.json --state.fork Byzantium
    test_cmd.go:261: (stderr:6) INFO [06-25|16:52:13.126] rejected tx                              index=1 hash=0557ba..18d673 from=0x8A8eAFb1cf62BfBeb1741769DAE1a9dd47996192 error="nonce too low: address 0x8A8eAFb1cf62BfBeb1741769DAE1a9dd47996192, tx: 0 state: 1"
    test_cmd.go:261: (stderr:6) INFO [06-25|16:52:13.126] Trie dumping started                     root=84208a..ae4e13
    test_cmd.go:261: (stderr:6) INFO [06-25|16:52:13.126] Trie dumping complete                    accounts=3 elapsed="60.108µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/3/alloc.json --input.txs ./testdata/3/txs.json --input.env ./testdata/3/env.json --state.fork Berlin
    test_cmd.go:261: (stderr:9) INFO [06-25|16:52:13.152] Trie dumping started                     root=b7341d..857ea1
    test_cmd.go:261: (stderr:9) INFO [06-25|16:52:13.153] Trie dumping complete                    accounts=3 elapsed="90.49µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/4/alloc.json --input.txs ./testdata/4/txs.json --input.env ./testdata/4/env.json --state.fork Berlin
    test_cmd.go:261: (stderr:12) ERROR(4): getHash(3) invoked, blockhash for that block not provided
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/5/alloc.json --input.txs ./testdata/5/txs.json --input.env ./testdata/5/env.json --state.fork Byzantium --state.reward 0x80
    test_cmd.go:261: (stderr:15) INFO [06-25|16:52:13.205] Trie dumping started                     root=a7312a..c1f393
    test_cmd.go:261: (stderr:15) INFO [06-25|16:52:13.205] Trie dumping complete                    accounts=3 elapsed="70.846µs"
    t8n_test.go:299: args: t8n --output.body stdout --output.result "" --output.alloc "" --input.alloc ./testdata/13/alloc.json --input.txs ./testdata/13/txs.json --input.env ./testdata/13/env.json --state.fork London
    test_cmd.go:261: (stderr:16) INFO [06-25|16:52:13.234] Trie dumping started                     root=e4b924..6aef61
    test_cmd.go:261: (stderr:16) INFO [06-25|16:52:13.234] Trie dumping complete                    accounts=3 elapsed="104.878µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/13/alloc.json --input.txs ./testdata/13/signed_txs.rlp --input.env ./testdata/13/env.json --state.fork London
    test_cmd.go:261: (stderr:18) INFO [06-25|16:52:13.256] Trie dumping started                     root=e4b924..6aef61
    test_cmd.go:261: (stderr:18) INFO [06-25|16:52:13.256] Trie dumping complete                    accounts=3 elapsed="61.682µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/14/alloc.json --input.txs ./testdata/14/txs.json --input.env ./testdata/14/env.json --state.fork London
    test_cmd.go:261: (stderr:19) INFO [06-25|16:52:13.277] Trie dumping started                     root=6f0588..7f4bdc
    test_cmd.go:261: (stderr:19) INFO [06-25|16:52:13.277] Trie dumping complete                    accounts=2 elapsed="65.13µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/14/alloc.json --input.txs ./testdata/14/txs.json --input.env ./testdata/14/env.uncles.json --state.fork London
    test_cmd.go:261: (stderr:20) INFO [06-25|16:52:13.298] Trie dumping started                     root=6f0588..7f4bdc
    test_cmd.go:261: (stderr:20) INFO [06-25|16:52:13.298] Trie dumping complete                    accounts=2 elapsed="54.479µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/14/alloc.json --input.txs ./testdata/14/txs.json --input.env ./testdata/14/env.uncles.json --state.fork Berlin
    test_cmd.go:261: (stderr:21) INFO [06-25|16:52:13.318] Trie dumping started                     root=6f0588..7f4bdc
    test_cmd.go:261: (stderr:21) INFO [06-25|16:52:13.318] Trie dumping complete                    accounts=2 elapsed="53.961µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/19/alloc.json --input.txs ./testdata/19/txs.json --input.env ./testdata/19/env.json --state.fork London
    test_cmd.go:261: (stderr:22) INFO [06-25|16:52:13.338] Trie dumping started                     root=6f0588..7f4bdc
    test_cmd.go:261: (stderr:22) INFO [06-25|16:52:13.338] Trie dumping complete                    accounts=2 elapsed="62.368µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/19/alloc.json --input.txs ./testdata/19/txs.json --input.env ./testdata/19/env.json --state.fork ArrowGlacier
    test_cmd.go:261: (stderr:23) INFO [06-25|16:52:13.359] Trie dumping started                     root=6f0588..7f4bdc
    test_cmd.go:261: (stderr:23) INFO [06-25|16:52:13.359] Trie dumping complete                    accounts=2 elapsed="62.569µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/19/alloc.json --input.txs ./testdata/19/txs.json --input.env ./testdata/19/env.json --state.fork GrayGlacier
    test_cmd.go:261: (stderr:24) INFO [06-25|16:52:13.379] Trie dumping started                     root=6f0588..7f4bdc
    test_cmd.go:261: (stderr:24) INFO [06-25|16:52:13.379] Trie dumping complete                    accounts=2 elapsed="66.893µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc "" --input.alloc ./testdata/23/alloc.json --input.txs ./testdata/23/txs.json --input.env ./testdata/23/env.json --state.fork Berlin
    test_cmd.go:261: (stderr:25) INFO [06-25|16:52:13.401] Trie dumping started                     root=653343..95154d
    test_cmd.go:261: (stderr:25) INFO [06-25|16:52:13.401] Trie dumping complete                    accounts=3 elapsed="56.934µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/24/alloc.json --input.txs ./testdata/24/txs.json --input.env ./testdata/24/env.json --state.fork Merge
    test_cmd.go:261: (stderr:26) INFO [06-25|16:52:13.421] Trie dumping started                     root=9e4224..043c15
    test_cmd.go:261: (stderr:26) INFO [06-25|16:52:13.421] Trie dumping complete                    accounts=3 elapsed="65.042µs"
    t8n_test.go:299: args: t8n --output.body "" --output.result "" --output.alloc "" --input.alloc ./testdata/24/alloc.json --input.txs ./testdata/24/txs.json --input.env ./testdata/24/env-missingrandom.json --state.fork Merge
    test_cmd.go:261: (stderr:27) ERROR(3): post-merge requires currentRandom to be defined in env
    t8n_test.go:299: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/25/alloc.json --input.txs ./testdata/25/txs.json --input.env ./testdata/25/env.json --state.fork Merge
    test_cmd.go:261: (stderr:28) panic: runtime error: invalid memory address or nil pointer dereference
    test_cmd.go:261: (stderr:28) [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5f5dbb]
    test_cmd.go:261: (stderr:28) goroutine 1 [running]:
    test_cmd.go:261: (stderr:28) math/big.(*Int).Cmp(0xc00057da40?, 0x0?)
    test_cmd.go:261: (stderr:28) 	/usr/local/go/src/math/big/int.go:381 +0x1b
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/core.(*StateTransition).preCheck(0xc00028ebb0)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/core/state_transition.go:323 +0x6f2
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/core.(*StateTransition).TransitionDb(0xc00028ebb0)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/core/state_transition.go:396 +0x51
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/core.ApplyMessage(0x152ada8?, 0x152adb0?, 0xc000579158?)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/core/state_transition.go:184 +0x57
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/cmd/evm/internal/t8ntool.(*Prestate).Apply(0xc0003ea2d0, {{0x0, 0x0}, 0x0, 0x0, {0x0, 0x0, 0x0}}, 0xc0004ad6c0, {0x1694ef8, ...}, ...)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/cmd/evm/internal/t8ntool/execution.go:243 +0x17cf
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/cmd/evm/internal/t8ntool.Transition(0xc000267300)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/cmd/evm/internal/t8ntool/transition.go:188 +0xb65
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/internal/flags.MigrateGlobalFlags.func2.1(0xc000267300)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/internal/flags/helpers.go:100 +0x34
    test_cmd.go:261: (stderr:28) github.com/urfave/cli/v2.(*Command).Run(0x1f59480, 0xc000267300, {0xc0003ea1e0, 0xf, 0xf})
    test_cmd.go:261: (stderr:28) 	/home/anna/go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/command.go:274 +0x93f
    test_cmd.go:261: (stderr:28) github.com/urfave/cli/v2.(*Command).Run(0xc00044eb00, 0xc000266040, {0xc000040100, 0x10, 0x10})
    test_cmd.go:261: (stderr:28) 	/home/anna/go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/command.go:267 +0xb7d
    test_cmd.go:261: (stderr:28) github.com/urfave/cli/v2.(*App).RunContext(
    test_cmd.go:261: (stderr:28) 0xc00016c5a0, {0x169aad0, 0x2166400}, {0xc000040100, 0x10, 0x10})
    test_cmd.go:261: (stderr:28) 	/home/anna/go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/app.go:332 +0x566
    test_cmd.go:261: (stderr:28) github.com/urfave/cli/v2.(*App).Run(...)
    test_cmd.go:261: (stderr:28) 	/home/anna/go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/app.go:309
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/cmd/evm.TestMain.func1()
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/cmd/evm/t8n_test.go:35 +0x45
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/internal/reexec.Init(...)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/internal/reexec/reexec.go:31
    test_cmd.go:261: (stderr:28) github.com/ethereum/go-ethereum/cmd/evm.TestMain(0xc00043e280)
    test_cmd.go:261: (stderr:28) 	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/cmd/evm/t8n_test.go:42 +0x67
    test_cmd.go:261: (stderr:28) main.main()
    test_cmd.go:261: (stderr:28) 	_testmain.go:53 +0x195
    t8n_test.go:312: test 16, file ./testdata/25/exp.json: json parsing failed: unexpected end of JSON input
FAIL
FAIL	github.com/ethereum/go-ethereum/cmd/evm	0.413s

```

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
gasprice: do not keep Oracle's lastBaseFee and lastMinGasTip empty

Initialize these values if London or NeoXBurn are not active yet.
Otherwise gas price related RPC handlers won't work properly with London
or NeoXBurn disabled.

Fix the following error in tests for `eth_gasPrice` handler
```
--- FAIL: TestEthClient (0.24s)
    --- FAIL: TestEthClient/StatusFunctions (0.00s)
        ethclient_test.go:503: unexpected error: method handler crashed
FAIL
FAIL	github.com/ethereum/go-ethereum/ethclient	0.258s

```

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva reviewed (commented):
Once #230 is merged, we need to adjust Policy contract documentation, it's out-of-date: https://github.com/bane-labs/go-ethereum/tree/bane-main/contracts#policy

AnnaShaleva committed:
gasprice: move policy-based minGasPrice retrieval our of NeoXBurn

Bug introduced by e427cda.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
t8ntool: extend the list of valid forks with NeoXBurn

Fix
https://github.com/bane-labs/go-ethereum/pull/230#issuecomment-2208760497,
t8ntool needs state.fork to be specified to create proper config, the
default value for this CLI flag is GrayGlacier.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
eth: return Policy's minGasTipCap from suggestTipCap

If minGasTipCap is available in Policy, then return this value from
SuggestGasTipCap. Ref.
https://github.com/bane-labs/go-ethereum/pull/230#issuecomment-2212888965.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
Merge pull request #230 from bane-labs/adjust-basefee

Adjust BaseFee burning mechanism

AnnaShaleva commented on core/txpool/validation.go:
```diff
@@ -203,9 +204,12 @@ func ValidateTransactionWithState(tx *types.Transaction, signer types.Signer, op
+	// For LegacyTx, GasFeeCap and GasPrice are equal, so checking GasTipCap and GasFeeCap is enough
+	var minGasTipCap = opts.State.GetState(systemcontracts.PolicyProxyHash, systemcontracts.GetMinGasTipCapStateHash()).Big()
+	var baseFee = opts.State.GetState(systemcontracts.PolicyProxyHash, systemcontracts.GetBaseFeeStateHash()).Big()
+	if math.BigMin(tx.GasTipCap(), new(big.Int).Sub(tx.GasFeeCap(), baseFee)).Cmp(minGasTipCap) < 0 {
```
@txhsl, please, update Policy contract documentation wrt this new functionality.

txhsl replied:
Ref #249. I also update it about #216 and #242.

songb2 commented:
@AnnaShaleva @chenquanyu @txhsl  I've tested metamask and t8ntool with the latest commit on this pr https://github.com/bane-labs/go-ethereum/pull/230. No more issues found.
